platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java

```diff
@@ -425,93 +456,186 @@ public DataFileReader endWriting() throws IOException {
+            } catch (final Exception z) {
+                throw new RuntimeException("Exception in HDHM.endWriting()", z);
             }
-            // close files session
-            dataFileReader = fileCollection.endWriting(0, numOfBuckets);
-            // we have updated all indexes so the data file can now be included in merges
-            dataFileReader.setFileCompleted();
         } else {
             dataFileReader = null;
         }
-
         // clear put cache
         oneTransactionsData = null;
         return dataFileReader;
     }
 
     /**
-     * Reads a bucket with a given index from disk, updates given keys in it, and puts the bucket to
-     * a queue. If an exception is thrown, it's put to the queue instead, so the number of {@code
-     * ReadBucketResult} objects in the queue is consistent.
-     *
-     * @param bucketIndex The bucket index
-     * @param keyUpdates Key/value updates to apply to the bucket
-     * @param queue The queue to put the bucket or exception to
+     * A task to read a bucket identified by the given idex from disk and apply a list of
+     * updates to the keys to it. The task has no dependencies, it's executed right after
+     * creation.
      */
-    private void readUpdateQueueBucket(
-            final int bucketIndex, final BucketMutation keyUpdates, final Queue<ReadBucketResult> queue) {
-        try {
-            // The bucket will be closed on the lifecycle thread
-            final Bucket bucket;
-            BufferedData bucketData = fileCollection.readDataItemUsingIndex(bucketIndexToBucketLocation, bucketIndex);
-            if (bucketData == null) {
-                // create a new bucket
-                bucket = bucketPool.getBucket();
-                bucket.setBucketIndex(bucketIndex);
+    private class ReadUpdateBucketTask extends AbstractTask {
+
+        // Bucket index
+        private final int bucketIndex;
+
+        // List of updates to apply to the bucket
+        private final BucketMutation keyUpdates;
+
+        ReadUpdateBucketTask(final ForkJoinPool pool, final int bucketIndex, final BucketMutation keyUpdates) {
+            super(pool, 0);
+            this.bucketIndex = bucketIndex;
+            this.keyUpdates = keyUpdates;
+        }
+
+        private void createAndScheduleStoreTask(final Bucket bucket) {
+            // Create a subsequent "store bucket" task for the bucket
+            final StoreBucketTask storeTask = new StoreBucketTask(getPool(), bucket);
+            // The last created "store bucket" task. storeTask above will be set as an
+            // output dependency for that task to make sure tasks are running only one at
+            // a time. See StoreBucketTask for details
+            final StoreBucketTask prevTask = lastStoreTask.getAndSet(storeTask);
+            if (prevTask != null) {
+                // This will trigger prevTask execution as soon as its prev task is complete
+                prevTask.setNext(storeTask);
```

Comment:
Minor: just for my education: since the prevTask already exists and may have started on a separate thread already, is it technically possible for the prevTask to finish before this `setNext` call is performed here? What happens in that case? Will the `storeTask` ever be executed then?

Reply:
No this is not possible. For a task to run, both its dependencies must be set. One of the dependencies is "next", which is set above. It means, prevTask may not be completed (or even running) here